libs/MotionKit/source/MotionKit.cpp

```diff
@@ -58,73 +52,58 @@ void MotionKit::startStabilisation()
 {
 	stop();
 
-	_imukit.start();
-	_imukit.setOrigin();
+	_target_angles = _imukit.getEulerAngles();
 
-	_target_not_reached		  = false;
-	_stabilisation_requested  = true;
-	_rotate_x_turns_requested = false;
+	auto on_euler_angles_rdy_callback = [this](const EulerAngles &euler_angles) {
+		processAngleForStabilization(euler_angles);
+	};
 
-	_event_loop.start();
+	_imukit.onEulerAnglesReady(on_euler_angles_rdy_callback);
 }
 
 // LCOV_EXCL_START - Dynamic behavior, involving motors and time.
-void MotionKit::run()
+void MotionKit::processAngleForRotation(const EulerAngles &angles, Rotation direction)
 {
-	auto last_yaw			= kReferenceAngle;
-	auto rotations_executed = 0;
+	auto must_stop = [&] { return !_rotate_x_turns_requested && !_target_not_reached; };
 
-	auto must_rotate = [&] { return _rotate_x_turns_requested && rotations_executed != _rotations_to_execute; };
+	if (must_stop()) {
+		stop();
 
-	auto check_complete_rotations_executed = [&](auto current_yaw) {
-		if (std::abs(last_yaw - current_yaw) >= 300.F) {
-			++rotations_executed;
+		if (_on_rotation_ended_callback) {
+			_on_rotation_ended_callback();
 		}
-	};
-
-	while (must_rotate()) {
-		auto [current_pitch, current_roll, current_yaw] = _imukit.getEulerAngles();
 
-		check_complete_rotations_executed(current_yaw);
-
-		rtos::ThisThread::sleep_for(70ms);
-		last_yaw = current_yaw;
+		return;
 	}
 
-	_rotate_x_turns_requested = false;
-	_rotations_to_execute	  = 0;
+	_angle_rotation_sum += _rotation_control.calculateYawRotation(_euler_angles_previous.yaw, angles.yaw);
 
-	while (_stabilisation_requested || _target_not_reached) {
-		auto [pitch, roll, yaw] = _imukit.getEulerAngles();
-		auto [speed, rotation]	= _pid.processPID(pitch, roll, yaw);
+	if (_rotate_x_turns_requested && _target_not_reached) {
+		auto speed = _rotation_control.processRotationAngle(_angle_rotation_target, _angle_rotation_sum);
```

Comment:
why do you need this to be here? 

the RotationControl should handle this logic, not MotionKit

Reply:
the whole function should be in RotationControl

you push the raw yaw value as input and get speed + direction as output

(before we move all that to the control class completely)